deployer/src/runtime_manager.rs

```diff
@@ -0,0 +1,109 @@
+use std::{path::PathBuf, sync::Arc};
+
+use shuttle_proto::runtime::{self, runtime_client::RuntimeClient, SubscribeLogsRequest};
+use tokio::sync::Mutex;
+use tonic::transport::Channel;
+
+use crate::deployment::deploy_layer;
+
+#[derive(Clone)]
+pub struct RuntimeManager {
+    legacy: Option<RuntimeClient<Channel>>,
+    next: Option<RuntimeClient<Channel>>,
+    binary_bytes: Vec<u8>,
+    artifacts_path: PathBuf,
+    provisioner_address: String,
+    log_sender: crossbeam_channel::Sender<deploy_layer::Log>,
+}
+
+impl RuntimeManager {
+    pub fn new(
+        binary_bytes: &[u8],
+        artifacts_path: PathBuf,
+        provisioner_address: String,
+        log_sender: crossbeam_channel::Sender<deploy_layer::Log>,
+    ) -> Arc<Mutex<Self>> {
+        Arc::new(Mutex::new(Self {
+            legacy: None,
+            next: None,
+            binary_bytes: binary_bytes.to_vec(),
+            artifacts_path,
+            provisioner_address,
+            log_sender,
+        }))
+    }
+
+    pub async fn get_runtime_client(&mut self, is_next: bool) -> RuntimeClient<Channel> {
+        if is_next {
+            self.get_next_runtime_client().await
+        } else {
+            self.get_legacy_runtime_client().await
+        }
+    }
+
+    async fn get_legacy_runtime_client(&mut self) -> RuntimeClient<Channel> {
```

Comment:
It would be nice to deduplicate these, but it might not be so easy. Something for a future PR maybe.

Reply:
I thought about it but did not see an easy solution, let me try another swing though